File: examples/go/impl.go
package main

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/machanirobotics/protoc-mcp-gen/examples/proto/generated/go/todo/todopbv1"
	"google.golang.org/protobuf/types/known/emptypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// Compile-time checks: todoServer implements both the MCP and gRPC server interfaces.
var _ todopbv1.TodoServiceMCPServer = (*todoServer)(nil)
var _ todopbv1.TodoServiceServer = (*todoServer)(nil)

// todoServer is an in-memory TodoService implementation.
type todoServer struct {
	todopbv1.UnimplementedTodoServiceServer
	mu    sync.RWMutex
	todos map[string]*todopbv1.Todo
}

func newTodoServer() *todoServer {
	return &todoServer{todos: make(map[string]*todopbv1.Todo)}
}

func (s *todoServer) CreateTodo(_ context.Context, req *todopbv1.CreateTodoRequest) (*todopbv1.Todo, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	name := fmt.Sprintf("%s/todos/%s", req.GetParent(), req.GetTodoId())
	now := timestamppb.New(time.Now())

	todo := req.GetTodo()
	if todo == nil {
		todo = &todopbv1.Todo{}
	}
	todo.Name = name
	todo.CreateTime = now
	todo.UpdateTime = now

	s.todos[name] = todo
	return todo, nil
}

func (s *todoServer) GetTodo(_ context.Context, req *todopbv1.GetTodoRequest) (*todopbv1.Todo, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	todo, ok := s.todos[req.GetName()]
	if !ok {
		return nil, fmt.Errorf("todo %q not found", req.GetName())
	}
	return todo, nil
}

func (s *todoServer) ListTodos(_ context.Context, req *todopbv1.ListTodosRequest) (*todopbv1.ListTodosResponse, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var result []*todopbv1.Todo
	for _, t := range s.todos {
		result = append(result, t)
	}
	return &todopbv1.ListTodosResponse{Todos: result}, nil
}

func (s *todoServer) UpdateTodo(_ context.Context, req *todopbv1.UpdateTodoRequest) (*todopbv1.Todo, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	existing, ok := s.todos[req.GetTodo().GetName()]
	if !ok {
		return nil, fmt.Errorf("todo %q not found", req.GetTodo().GetName())
	}

	t := req.GetTodo()
	if t.Title != "" {
		existing.Title = t.Title
	}
	if t.Description != "" {
		existing.Description = t.Description
	}
	existing.Completed = t.Completed
	existing.Priority = t.Priority
	existing.UpdateTime = timestamppb.New(time.Now())

	return existing, nil
}

func (s *todoServer) DeleteTodo(_ context.Context, req *todopbv1.DeleteTodoRequest) (*emptypb.Empty, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, ok := s.todos[req.GetName()]; !ok {
		return nil, fmt.Errorf("todo %q not found", req.GetName())
	}
	delete(s.todos, req.GetName())
	return &emptypb.Empty{}, nil
}

File: examples/python/impl.py
"""In-memory TodoService implementation."""

from __future__ import annotations

from datetime import datetime, timezone

from google.protobuf.empty_pb2 import Empty
from google.protobuf.timestamp_pb2 import Timestamp

from todo.v1 import todo_pb2


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _now() -> Timestamp:
    """Return the current UTC time as a protobuf Timestamp."""
    ts = Timestamp()
    ts.FromDatetime(datetime.now(timezone.utc))
    return ts


# ---------------------------------------------------------------------------
# Implementation
# ---------------------------------------------------------------------------

class TodoServer:
    """Implements the TodoServiceMCPServer protocol with an in-memory store."""

    def __init__(self) -> None:
        self._todos: dict[str, todo_pb2.Todo] = {}

    async def create_todo(
        self, request: todo_pb2.CreateTodoRequest
    ) -> todo_pb2.Todo:
        name = f"{request.parent}/todos/{request.todo_id}"
        now = _now()

        todo = todo_pb2.Todo()
        if request.HasField("todo"):
            todo.CopyFrom(request.todo)
        todo.name = name
        todo.create_time.CopyFrom(now)
        todo.update_time.CopyFrom(now)

        self._todos[name] = todo
        return todo

    async def get_todo(
        self, request: todo_pb2.GetTodoRequest
    ) -> todo_pb2.Todo:
        todo = self._todos.get(request.name)
        if todo is None:
            raise ValueError(f"todo {request.name!r} not found")
        return todo

    async def list_todos(
        self, request: todo_pb2.ListTodosRequest
    ) -> todo_pb2.ListTodosResponse:
        resp = todo_pb2.ListTodosResponse()
        for todo in self._todos.values():
            resp.todos.append(todo)
        return resp

    async def update_todo(
        self, request: todo_pb2.UpdateTodoRequest
    ) -> todo_pb2.Todo:
        existing = self._todos.get(request.todo.name)
        if existing is None:
            raise ValueError(f"todo {request.todo.name!r} not found")

        if request.todo.title:
            existing.title = request.todo.title
        if request.todo.description:
            existing.description = request.todo.description
        existing.completed = request.todo.completed
        existing.priority = request.todo.priority
        existing.update_time.CopyFrom(_now())

        return existing

    async def delete_todo(
        self, request: todo_pb2.DeleteTodoRequest
    ) -> Empty:
        if request.name not in self._todos:
            raise ValueError(f"todo {request.name!r} not found")
        del self._todos[request.name]
        return Empty()
